kore/src/Kore/Attribute/Synthetic.hs

```diff
@@ -6,45 +6,104 @@ License     : NCSA
  -}
 
 module Kore.Attribute.Synthetic
-    ( synthesize
+    ( Synthetic (..)
+    , resynthesize, resynthesizeAux
+    , synthesize, synthesizeAux
     ) where
 
 import           Control.Comonad.Trans.Cofree
                  ( CofreeF (..) )
 import qualified Control.Comonad.Trans.Cofree as Cofree
+import           Data.Functor.Const
 import           Data.Functor.Foldable
                  ( Base, Corecursive, Recursive )
 import qualified Data.Functor.Foldable as Recursive
 
+{- | @Synthetic@ is the class of synthetic attribute types @syn@.
+
+@Synthetic base syn@ allows synthesizing @syn@ given a @'Cofree' base@ tree;
+that is, a 'Cofree' tree with branching described by a @'Functor' base@.
+
+ -}
+class Functor base => Synthetic base syn where
+    -- | @synthetic@ is the @base@-algebra for synthesizing the attribute @syn@.
+    synthetic :: base syn -> syn
+
+instance Synthetic (Const a) a where
+    synthetic (Const a) = a
+    {-# INLINE synthetic #-}
+
 {- | @/synthesize/@ attribute @b@ bottom-up along a tree @s@.
 
 @synthesize@ is a generalization of 'Data.List.scanr' to trees: Given a tree @s@
-with attributes @a@ along the nodes, @synthesize@ produces a tree @t@ with
-attributes @b@ along the nodes using the given @(Base s)@-algebra from the
-bottom up. The algebra's argument is the original @a@-attribute of a node and
-the @b@-attributes of all children.
+with attributes @inh@ along the nodes, @synthesize@ produces a tree @t@ with
+attributes @syn@ along the nodes using the given @('Base' s)@-algebra from the
+bottom up.
 
 See also:
 <https://en.wikipedia.org/wiki/Attribute_grammar#Synthesized_attributes>
 
  -}
-synthesize
+resynthesize
```

Comment:
One more thing: Does this synthesize stuff work with things that contain concrete stuff, e.g. set builtins?

Reply:
Right now, the instance for builtins ignores the concrete stuff, but we can fix that. I didn't think about it because the concrete parts don't usually change during execution.